### PR TITLE
fix(dropdown): Dropdown menu support transmit ref

### LINKED ISSUE #42979: [Dropdown] focus doesn't go to the menu when using autoFocus prop
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/antd-reproduction-template-forked-wl3dkl?file=/index.js)

### Steps to reproduce

1) Click on the three dot icon to open the menu
2) Try to navigate using the keyboard, use the arrow keys to select a menu item

### What is expected?

The focus shifts to the menu and I can select a menu item

### What is actually happening?

The focus doesn't shift to the menu and I can't select a menu item

| Environment | Info |
| --- | --- |
| antd | 5.6.1 |
| React | 18.2.0 |
| System | Pop!_OS 22.04 LTS |
| Browser | Chrome 113.0.5672.126 |

---
This problem first appeared in version 5.5.0, I could not reproduce it in 5.4.7.
https://codesandbox.io/s/antd-reproduction-template-forked-wky7gz?file=/index.js

### PR DESCRIPTION
**Pull request by Yuiai01**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/42979
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Dropdown menu support transmit ref           |
| 🇨🇳 Chinese | 下拉菜单支持传递 `ref`           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 2070464</samp>

Fixed a bug and improved the spacing of dropdown menus by moving `NoCompactStyle` from `dropdown.tsx` to `OverrideProvider` component. Refactored `OverrideProvider` to use `React.forwardRef` and `NoCompactStyle`.

### <samp>🤖 Generated by Copilot at 2070464</samp>

* Remove `NoCompactStyle` import and wrapper from `dropdown.tsx` to fix spacing issues ([link](https://github.com/ant-design/ant-design/pull/43002/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beL17), [link](https://github.com/ant-design/ant-design/pull/43002/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beL277-R276))
* Add `NoCompactStyle` import and wrapper to `OverrideContext.tsx` to ensure consistent spacing and behavior of dropdown menu ([link](https://github.com/ant-design/ant-design/pull/43002/files?diff=unified&w=0#diff-fab289ecf381929ccfd1ab5e53c7170d3e249910d5774f13f2a76eca26bb6630R2), [link](https://github.com/ant-design/ant-design/pull/43002/files?diff=unified&w=0#diff-fab289ecf381929ccfd1ab5e53c7170d3e249910d5774f13f2a76eca26bb6630L35-R42))
* Refactor `OverrideProvider` component to use `React.forwardRef` and `React.cloneElement` to pass ref and context to children ([link](https://github.com/ant-design/ant-design/pull/43002/files?diff=unified&w=0#diff-fab289ecf381929ccfd1ab5e53c7170d3e249910d5774f13f2a76eca26bb6630L17-R21))

### CODE REVIEW

**Comment by afc163** [comment]
补个用例

**Comment by Yuiai01** [comment]
> 补个用例

好，但是感觉目前这种方法会存在问题，我再康康

**Comment by Yuiai01** [comment]
> 覆盖率掉了

这种的该咋处理啊

**Comment by kiner-tang** [comment]
> > 覆盖率掉了
> 
> 这种的该咋处理啊

点进去看一下哪一行没有没覆盖到，然后增加测试用例覆盖就好了

**Comment by kiner-tang** [comment]
> > > 覆盖率掉了
> > 
> > 
> > 这种的该咋处理啊
> 
> 点进去看一下哪一行没有没覆盖到，然后增加测试用例覆盖就好了

https://app.codecov.io/gh/ant-design/ant-design/pull/43002?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=ant-design
image

**Comment by Yuiai01** [comment]
> > > > 覆盖率掉了
> > > 
> > > 
> > > 这种的该咋处理啊
> > 
> > 
> > 点进去看一下哪一行没有没覆盖到，然后增加测试用例覆盖就好了
> 
> https://app.codecov.io/gh/ant-design/ant-design/pull/43002?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=ant-design image

嗯嗯 刚刚点进去看到了 感谢~

File: components/menu/OverrideContext.tsx [hunk 1]
```diff
@@ -32,8 +34,12 @@ export const OverrideProvider: React.FC<OverrideContextProps & { children: React
     ],
   );
 
-  return <OverrideContext.Provider value={context}>{children}</OverrideContext.Provider>;
-};
+  return (
+    <OverrideContext.Provider value={context}>
+      <NoCompactStyle>{React.cloneElement(children as React.ReactElement, { ref })}</NoCompactStyle>
```
Comment (afc163) [inline components/menu/OverrideContext.tsx:39]: 为啥要加 NoCompactStyle ？

Comment (Yuiai01) [reply components/menu/OverrideContext.tsx:39]: 因为之前外部是嵌套了一个 <NoCompactStyle>，为了保持一致性，就给一起挪过来了。

**Review by afc163 — APPROVED** [review]

**Pull request merged**